src/qe/mbp/mbp_basic_tg.cpp

```diff
@@ -43,32 +43,106 @@ struct mbp_basic_tg::impl {
         expr_ref nterm(m);
         bool progress = false;
         TRACE("mbp_tg", tout << "Iterating over terms of tg";);
         // Not resetting terms because get_terms calls resize on terms
         m_tg.get_terms(terms, false);
-        for (expr* term : terms) {
-            if (is_seen(term))
-                continue;
-            if (m.is_ite(term, c, th, el)) {
+        for (expr *term : terms) {
+            if (is_seen(term)) continue;
+            if (m.is_ite(term, c, th, el) && should_split(c)) {
                 mark_seen(term);
                 progress = true;
                 if (m_mdl.is_true(c)) {
                     m_tg.add_lit(c);
                     m_tg.add_eq(term, th);
-                }
-                else {
+                } else {
                     nterm = mk_not(m, c);
                     m_tg.add_lit(nterm);
                     m_tg.add_eq(term, el);
                 }
             }
+            if (m.is_implies(term, c, th)) {
+                if (is_true(th) || is_false(c)) {
+                    mark_seen(term);
+                    progress = true;
+                    if (is_true(th))
+                        m_tg.add_lit(th);
+                    else if (is_false(c))
+                        m_tg.add_lit(c);
+                    m_tg.add_eq(term, m.mk_true());
+                } else if (is_true(c) && is_false(th)) {
+                    mark_seen(term);
+                    progress = true;
+                    m_tg.add_eq(term, m.mk_false());
+                }
+            }
+            if (m.is_or(term) || m.is_and(term)) {
+                bool is_or = m.is_or(term);
+                app *c = to_app(term);
+                bool t = is_or ? any_of(*c, is_true) : all_of(*c, is_true);
+                bool f = is_or ? all_of(*c, is_false) : all_of(*c, is_false);
+                if (t || f) {
+                    mark_seen(term);
+                    progress = true;
+                    m_tg.add_eq(term, t ? m.mk_true() : m.mk_false());
+                    if (f) {
+                        for (auto a : *c) {
+                            if (is_false(a)) {
+                                m_tg.add_lit(mk_not(m, a));
+                                if (!is_or) break;
+                            }
+                        }
+                    } else {
+                        for (auto a : *c) {
+                            if (is_true(a)) {
+                                m_tg.add_lit(a);
+                                if (is_or) break;
+                            }
+                        }
+                    }
+                }
+            }
+            if (m_use_mdl && m.is_distinct(term)) {
+                mark_seen(term);
+                progress = true;
+                bool eq = false;
+                app *c = to_app(term);
+                for (auto a1 : *c) {
+                    for (auto a2 : *c) {
+                        if (a1 == a2) continue;
```

Comment:
It is OK to have single line statement after if on a separate line. The code base isn't consistent about this but tends to mostly breaking a line after if.